### ISSUE TITLE
[SR] Enforce some basic password constraints: no username, no email

### ISSUE THREAD

**Issue by herbdool** [backdrop/backdrop-issues#4265]
**Description of the need**
Currently Backdrop (and D7) allow anything to be entered in the password. I think that's okay for the most part (contrib modules like Password Policy (d7-only?) can provide more constraints) but I do feel that there are a couple obvious constraints that should be enforced: **username** and **email address**.

**Proposed solution**
Don't allow users to enter username or email in their password.

---
~PR by @indigoxela: https://github.com/backdrop/backdrop/pull/3053~
~PR by @jenlamptonhttps://github.com/backdrop/backdrop/pull/3281~ (separated UI for pw length)
~PR by @herbdool https://github.com/backdrop/backdrop/pull/3215~
PR by @jenlampton https://github.com/backdrop/backdrop/pull/3282 (removed UI for pw length)

**Comment by indigoxela** [comment]
There's a Backdrop module for that: https://github.com/backdrop-contrib/passphrase_policy (didn't use it yet)

But maybe it's really smarter to prevent users from setting their username or email as password (which is a horrible idea anyway).

What do people think?

**Comment by herbdool** [comment]
I agree that the passphrase_policy module is useful. Though I think that we can still provide some minimum safeguards in core.

I found this article quite helpful in what are some basics that we should enforce for all users: https://blog.codinghorror.com/password-rules-are-bullshit/

**Comment by herbdool** [comment]
Hm, just found out that `Backdrop.evaluatePasswordStrength` actually includes some of this checking: entropy, username as password, etc, based on https://github.com/dropbox/zxcvbn. It will just mark the password as poor but still allow people to save it.

**Comment by indigoxela** [comment]
The question is: why don't we already enforce at least a minimum length?

Currently even "0" is considered a password. Found [that](https://github.com/backdrop/backdrop/blob/9ed127933df9994ec3e96f9110520359451a183d/core/modules/user/tests/user.test#L1791) in the tests... , but I don't understand the reason for this.

The next question is: how much could get enforced without frustrating users?

- Not username
- Not email
- Not site name
- Not domain
- Minimum length (how much?)
- ...?

Or is this already too much? From a security perspective, this isn't enough, though.

I usually don't use modules like passphrase_policy/password_policy because the risk of frustrating users is big. On the other hand zero validation doesn't seem right to me either.

**Comment by alanmels** [comment]
I thought to advocate the Passphrase Policy module, because it works great on couple of our test setups. However, after reading the last argument of @indigoxela that some users could get frustrated about the strictness of password policies, I also agree that some minimum measures must be enforced in core, leaving stricter policies to contrib.

**Comment by stpaultim** [comment]
I have very mixed feelings about this. I think it's critical that users have the ability to add a passphrase policy and the contrib modules seems to exist that does this. 

During development, it's nice to just set username and password to the same thing. I don't do it myself, but I see tools like Lando do this and it's convenient during local development. 

I could see the value of a config option in core that forces some minimal password requirements. I'd prefer the ability to turn that off for development, but as long as the requirements are not too complicated I guess I don't feel strongly about it. 

As long as the module is available, I don't see this is a big problem either.

**Comment by Graham-72** [comment]
I agree that we need some minimum measures in core like

- at least 5 characters (any)

- but not username nor email.

Anything more should have an admin checkbox to apply it.

**Comment by indigoxela** [comment]
As @herbdool already mentioned in a previous comment, we already have smart password strength evaluation (in [user.js](https://github.com/backdrop/backdrop/blob/9ed127933df9994ec3e96f9110520359451a183d/core/modules/user/js/user.js#L146))

We probably only need to add the same logic to a php validate function.

A quick idea is to add a setting to /admin/config/people/settings like "Password policy".
Probably some form fields:

- Minimum length (integer, default 5?)
- Prevent user name (checkbox, unchecked?)
- Prevent account email (checkbox, unchecked?)
- Enforce strong password (checkbox, unchecked)

The latter would enforce a calculated value above - let's say - 80.
(But... some of these fields/settings seem conflicting to me. And maybe that's too technical for admins)

What do you think?

**Comment by Graham-72** [comment]
:+1: I like this proposal.

**Comment by indigoxela** [comment]
I see consensus, that core should have a _minimal_ password constraint feature (configurable).

I've created a PR for further discussion.
This needs improvement. Some help text for the admin form, for instance.

![screenshot-pass-constraints](https://user-images.githubusercontent.com/842066/73286524-fa486f00-41f7-11ea-966b-da0b0ab59bdf.png)

Are there any opinions yet?

**Comment by stpaultim** [comment]
Good work.

**Comment by olafgrabienski** [comment]
> Are there any opinions yet?

I like the option to have these basic password constraints in core.

> During development, it's nice to just set username and password to the same thing. I don't do it myself, but I see tools like Lando do this and it's convenient during local development.

If Lando or other environments do this, they'd work only if the default for `pass_not_username` is `FALSE`. We could do so and put a message on the status report page with a link to change the setting.

**Comment by indigoxela** [comment]
> If Lando or other environments do this, they'd work only if the default for pass_not_username is FALSE

This is one of the things to evaluate. To my knowledge the install script uses different form items (not user_account_form), so the constraints don't apply there.
And I think (to be evaluated!), drush uses user_save directly, which also means that the constraints don't apply.
Of course, we don't want to break scripted installing.

**Comment by indigoxela** [comment]
Tests are passing now.

To move on, I'd like to get some help regarding text/wording of help texts. I already learned that many people here create much better help texts than I do :wink: 

What information is helpful for admins when setting constraints? Should we explain, what these settings mean?

Another thought was: users nowadays almost expect that they have to input something cryptic with numbers/lowercase/uppercase/symbols...

But actually a long string (sentence) is better than a short complex password.
What if we put something like "Long and simple is better than short and hard" at the bottom of the profile/reset form help text? (Just an idea...)

Another task: check that automated installs and user creation isn't blocked by this new feature. Should we create a todo-list for that?

**Comment by klonos** [comment]
Not sure about this (although I do not necessarily object). ...I mean, we do have contrib that allows enforcing password rules/policy, and I believe that communicating that the password strength is enough for core.

What we need to fix is this:

![Screen Shot 2020-01-30 at 6 11 13 am](https://user-images.githubusercontent.com/2423362/73389022-6d043800-4327-11ea-8e12-4ef6e02cda4a.png)

^^ I have copy-pasted the email in the password field. I think that we should change this so that in that case the strength indicator is "poor".

If we decide that this is a core-worthy feature after all, then I would like to propose that we make it opt-in, instead of enforcing things. So let's keep whatever defaults we decide are best, but let's do something like this:

![Screen Shot 2020-01-30 at 6 25 49 am](https://user-images.githubusercontent.com/2423362/73390087-84442500-4329-11ea-8f60-f170c2230693.png)

**Comment by klonos** [comment]
... @indigoxela while at it, I would be really grateful if you also removed the periods from the checkboxes and radio options in that form ...would be a shame to raise a separate issue and file a PR only for that 😉

**Comment by klonos** [comment]
...and for reference, here's how password setting works in WordPress:

1. Password is automatically generated, and hidden from the admin (I absolutely love this feature btw):

    ![Screen Shot 2020-01-30 at 6 37 34 am](https://user-images.githubusercontent.com/2423362/73391037-58c23a00-432b-11ea-80d3-3fc454716683.png)

2. Password can be shown, by clicking the respective button:

    ![Screen Shot 2020-01-30 at 6 38 09 am](https://user-images.githubusercontent.com/2423362/73391207-9b841200-432b-11ea-8a22-131b388a69af.png)

3. If you manually type a weak password, there is a confirmation checkbox:

    ![Screen Shot 2020-01-30 at 6 38 20 am](https://user-images.githubusercontent.com/2423362/73391311-cd957400-432b-11ea-8da0-eb5c21b59350.png)

**Comment by docwilmot** [comment]
TBH I lean towards "there's a contrib solution" for this myself.

**Comment by Graham-72** [comment]
I rather like this one (copied from above) and in core please.
![image](https://user-images.githubusercontent.com/10559274/73397065-977bd400-42da-11ea-871d-cd1ac9a64442.png)

**Comment by stpaultim** [comment]
I was going to suggest making it possible to opt out and then I realized that if set the minimum length to "1" and unselect both boxes, you have essentially opted out. After all, it's impossible to have a password with less than one character (or am I missing something). 

Having said that, I'm not against adding the "enforce constraints" radio button. I'm just noting that it MIGHT not be necessary (just convenient). Also, it does change the default, which I like. 

IF we go with @klonos suggestion. I would recommend:

TITLE: Password Enforcement

O - No constraints - include password strength indicator 
O - Enforce minimum constraints on user passwords

(If box two is checked, then show the "constraint" options.) 

NOTE: Thanks for work on this. I don't feel strongly about these changes, just offering them as ideas/suggestions.

**Comment by klonos** [comment]
>(If box two is checked, then show the "constraint" options.)

These options should always be shown. They'd effect the "poor" password strength indicator - even if no enforcement.

**Comment by indigoxela** [comment]
> we do have contrib that allows enforcing password rules/policy

Yes, and what we already discussed here in this issue is, that many of us find it wrong to have _zero_ validation in core.
Based on this discussion and consensus, I've created a PR.

Currently there's only a single Backdrop contrib module for password constraints, and that one is _way_ too strict and has a weird UI.
So Backdrop users only have the choice between no validation at all and frustratingly strict validation without ability to configure anything.

> I would be really grateful if you also removed the periods from the checkboxes and radio options in that form

Sure... you mean the help text list, right?

I already [asked for help](https://github.com/backdrop/backdrop-issues/issues/4265#issuecomment-579690857) re help text and admin form description. :wink:

**Comment by herbdool** [comment]
Yeah, I think it's wrong to leave all password validation to contrib. Only a few users will take advantage of it. The 80% will assume that Backdrop is going to keep their site secure. Having some obvious password validation (that match the JS suggestions) is an easy win and will help everyone.

**Comment by indigoxela** [comment]
I also like @klonos' suggestion to make it fully opt-in (with radios).
Then there's no change in core behavior until an admin decides to use this setting.

* Only indicate poor password strength
* Enforce minimum constraints on user passwords

I find @stpaultim's text suggestion for the second radio slightly better.

And I think it's better to toggle the constraint options visibility state based on these radios.
Then it's clearer, that these settings only apply, if constraints are enabled.

**Comment by herbdool** [comment]
@indigoxela perhaps existing sites don't have it enabled by default but new sites have it enabled.

**Comment by klonos** [comment]
> Sure... you mean the help text list, right?

Nope, I mean these:

![Screen Shot 2020-01-31 at 2 18 35 am](https://user-images.githubusercontent.com/2423362/73462820-7ac1c800-43d0-11ea-879b-096c6446541c.png)

![Screen Shot 2020-01-31 at 2 18 51 am](https://user-images.githubusercontent.com/2423362/73462875-8c0ad480-43d0-11ea-89f0-57db7da866e2.png)

AFAIK, radio and checkbox options do not have periods (even when they are phrases).

**Comment by klonos** [comment]
>And I think it's better to toggle the constraint options visibility state based on these radios.
>Then it's clearer, that these settings only apply, if constraints are enabled.

The way I see it, even when the admin selects to only indicate poor password strength, they still need to control which aspects would affect this. So the various options should be always visible to allow affecting the "threshold" of poor. In other words:

- the various controls would allow to specify what constitutes a poor password
- the radios would allow to specify what Backdrop does with poor passwords (indicate only - or prevent account creation)

**Comment by indigoxela** [comment]
> Nope, I mean these:

@klonos I see... I'm afraid, that's out of the scope of this issue. At least, I'd hesitate (and wouldn't recommend) to mix it with the password policy feature.

**Comment by klonos** [comment]
> I'm afraid, that's out of the scope ...

I know 😞 ...it'd just seem pointless to raise a separate issue and file a PR only for those. When I work on UX/UI issues, I alway fix small things like these if they are in the same form.

**Comment by indigoxela** [comment]
> even when the admin selects to only indicate poor password strength, they still need to control which aspects would affect this

Sorry, I don't get it yet...

The strength indicator is pure javascript and (at least currently) not in any way connected to these settings (which are for server side validation).
The logic behind that indicator is slightly different and more complex.

The three options that are currently used for server side validation are the ones that we were able to agree on as being "basic constraints". (Note the title of this issue) :wink:

**Comment by klonos** [comment]
Yeah, this is how I thought this feature would work:

- We currently have a password strength indicator, but this does not have any actual validation effect.
- This indicator is based only on https://github.com/dropbox/zxcvbn
- With this issue here, we introduce various additional factors that affect this indicator.
- We also introduce an option to either allow poor passwords, or prevent the user creation if password is poor.

**Comment by indigoxela** [comment]
@klonos I think you're over-complicating things a bit. I always saw the indicator and server side validation as completely independent from each other.

Besides the fact that Backdrop.evaluatePasswordStrength (user.js) also needs to take email address fields into account when calculating strength, I wouldn't change anything there.

**Comment by klonos** [comment]
Yeah, but see the disconnect between visual indication and validation error in this for example:

![Screen Shot 2020-01-30 at 6 11 13 am](https://user-images.githubusercontent.com/2423362/73389022-6d043800-4327-11ea-8e12-4ef6e02cda4a.png)

^^ this tells the user that the password they entered is "green", but if they try to save the form, they'll get a validation error (because the password same as email constraint is set). I would expect the indicator to reflect the fact that entering the email as password is considered "poor".

**Comment by indigoxela** [comment]
@klonos ...please... read my previous comment. I already explained that.

**Comment by indigoxela** [comment]
As some of you were really baffled by the mismatch of js indicator and server side validation, I've fixed that now. Email is also considered now for indication.

I also added the radios to opt-in and opt-out.

According to @herbdool's idea to turn it on by default for fresh installs, but keep it off for existing ones, I tried that now.
This might need further discussion, though. I don't know if Backdrop does things like that anyway (different settings for existing and new).

**Comment by docwilmot** [comment]
@indigoxela good stuff. 
Would it be possible to have a different password strength method for passwords that would *fail* validation? Instead of just saying "poor", say "not allowed" or such? I do agree that it seems contradictory to say the password is just poor when we know well that it wont even be allowed.
Not absolutely necessary, I suppose, but seems would be more complete, and the code doesnt look like it would be much.

**Comment by indigoxela** [comment]
> Would it be possible to have a different password strength method for passwords that would fail validation?

That's an interesting idea...
Do I get this right, you mean to enhance the indicator to show "not allowed" if:

- Password is username and that's forbidden by admin setting
- Password is email and that's forbidden by admin setting
- Password shorter than admin setting

This might be easy, or maybe it's not - I'll have a look ASAP.

**Comment by docwilmot** [comment]
I actually meant to type "message", not "method" but yes you got the exact idea.

**Comment by klonos** [comment]
Brilliant!! Thanks @indigoxela and @docwilmot 👍

**Comment by klonos** [comment]
I've done a quick test in the PR sandbox, and everything works as expected 👍 ...I have made some UI text suggestions (mainly avoiding using the word "please").

**Comment by indigoxela** [comment]
@klonos many thanks for your help regarding wording.

The js strength indicator now considers constraint settings as suggested.

Hm... to be honest, I'm not convinced yet. This is somewhat strange, especially regarding minlength. But try it yourself.

**Comment by indigoxela** [comment]
The PR is currently conflicting - that was expected, I'll fix that later.

To describe why I think that we shouldn't show "not allowed" when the password is too short:

When not copy/pasting the password, but actually typing, the "not allowed" appears when starting to type.

That's a WTF moment for users, they might think "Hey, what did I do wrong? I only started typing!".

Instead of "not allowed" maybe we should show a very short red bar. Or are there better suggestions?

**Comment by indigoxela** [comment]
Based on my findings (see my previous comment), I've updated the PR.

Please test and review, feedback is welcome.

**Comment by stpaultim** [comment]
I think it's pretty good. I wasn't convinced it was necessary, but I understand the reason. My personal preference would be to have the enforcement off by default - but that is probably the developer speaking, not the site admin. 

One last idea (although this could be a follow-up issue or it could be rubbish). Would it be worth having something on the site status page for password enforcement. Just a reminder to a site admin as their checking their site status to let them know if password enforcement is on or not?

I recognize that real estate is precious on that page, but given the security implications would this be worth it? (I do not feel strongly about this, just throwing it out there)

**Comment by indigoxela** [comment]
> My personal preference would be to have the enforcement off by default

That's one of the things that are pending a final decision. Currently in my PR it's off for existing sites (when updating), but on for new sites. That's a core team decision, though.

> Would it be worth having something on the site status page for password enforcement...

That's a good point. What should we describe there? Would that be a hint or a warning? (Rather not a warning, as probably the password constraints should be fully optional.)

**Comment by klonos** [comment]
>My personal preference would be to have the enforcement off by default

I'm neither hot nor cold about this ATM.

> Would it be worth having something on the site status page for password enforcement...

I think that this is a better candidate for the dedicated "Security" section (#3624). I would then expect the status page to only hold a "Security summary" indication, so things do not become cluttered. What do you think @stpaultim ?

@indigoxela I would expect the "not allowed" indicator text color to be red (same as poor), but here's what it currently looks like:

![Screen Shot 2020-02-25 at 5 26 01 am](https://user-images.githubusercontent.com/2423362/75180318-0f310780-5790-11ea-9167-9619a9e66f2f.png)

...here's an idea:

![Screen Shot 2020-02-25 at 6 05 08 am](https://user-images.githubusercontent.com/2423362/75182612-e0696000-5794-11ea-8a5d-5b1af271a442.png)

Denoting the "offending" restriction in the password field description with a red color + all "passing" restrictions with green (and ✓/✗ instead of bullets), the user will have a better idea of what they may be doing wrong.

**Comment by stpaultim** [comment]
> I think that this is a better candidate for the dedicated "Security" section (#3624). I would then expect the status page to only hold a "Security summary" indication, so things do not become cluttered. What do you think @stpaultim ?

I would agree. But, I could see adding it to the status page until there is a security page to put it on. Unless, we're confident that the security page is coming soon. ;-)

**Comment by klonos** [comment]
@indigoxela one more thing: I think that `Password strength: not allowed` does not make sense; how about something like `Password strength: weak (not allowed)` ...so the `(not allowed)` thing gets appended only if the restrictions are enforced.

**Comment by indigoxela** [comment]
> I think that this is a better candidate for the dedicated "Security" section (#3624)

I agree, a security section seems more suitable for information like that. Is there any time schedule yet for that new feature?

Regarding the colors: currently there are no tags/wrappers for styling these text parts yet.

Regarding replacing the list bullets with utf-8 chars in the description text (item list, translatable): is there any example for an existing help text, that does it that way?

`Password strength: weak (not allowed)` ehm... @klonos are you sure?
I'm not a native speaker, but hat seems a little weird to me. Especially the combination of "weak" and "not allowed". I'd suppose, something is either _weak_ or _not allowed_ (not usable all), but to my understanding it can't be both.

**Comment by indigoxela** [comment]
Regarding the wording, we currently struggle with, I had an idea:
What about "**too weak**" instead of "not allowed"?

That fits the label "Password strength" and still denotes the password as being not strong enough yet. Or are there better ideas?

**Comment by docwilmot** [comment]
I think that's perfect. Too weak makes sense let's not complicate this further. 

Also note this could have been a contributed module. But we decided that it weakened Backdrop not to have it in core. So having it disabled by default makes no sense. We might as well have said if you want it install a module, now we're suggesting if you want it you'll have to turn it in. I vote, in summary, for on by default. No one will be turning this off

**Comment by klonos** [comment]
> Regarding the colors: currently there are no tags/wrappers for styling these text parts yet.

> What about "too weak" instead of "not allowed"?

Yes! ...that's makes a lot more sense. Thanks 👍 

> ...having it disabled by default makes no sense.

Yeah, I'm ambivalent re this. ...initially I thought best to leave existing sites alone, but now I see that this would not affect existing user accounts - only new ones being created. And sensible defaults wherever possible do make for better UX.

**Comment by indigoxela** [comment]
I realized that if we're using "too weak" as text, then we can show that for too short passwords too (for consistency). And I found out how to style the indicator text (red).

> ... GovCMS sites a lot, and these are using password_policy, which makes it so that each item in the list disappears as each requirement is fulfilled

Sure, that makes sense for password_policy, as they have a lot more requirements. We're only dealing with some basic constraints here, which can get disabled. So, to me this doesn't look like much help for users, actually. But I could be wrong.

Regarding the colored help text and the ✓/✗ as toggled list icons. - I have no clue, how this could get done. Maybe a follow-up issue/PR by someone else with better js-fu?

**Comment by docwilmot** [comment]
Looks good RTBC

**Comment by BWPanda** [comment]
I posted a review to the PR requesting some changes (mostly small coding standard stuff, but there was a query there about a possible API change...

Thanks for your work on this @indigoxela! If you need any help re. coding standards, check out https://serundeputy.io/automate-backdrop-coding-standards-your-projects

**Comment by indigoxela** [comment]
@BWPanda Many thanks for taking the time to review. I left some answers to your questions on my PR.

(I'm pretty sure the failed BootstrapPageCacheTestCase is unrelated)

Re coding standards: I do have phpcs installed. But unfortunately only for file user.module it finds 54 violations.
I'm striving to not add even more, but it's a bit harder if there are so many already.

**Comment by oadaeh** [comment]
Hello all. I've been following this issue from a distance, but now with the looming release, I'm taking a closer look at it.

While I'm likely to increase the complexity requirements for my own sites (because I do not believe that security should be sacrificed for user laziness and frustration (but I do believe communication and information goes a long way to reducing user frustration, as I demonstrate below)), I like what I see here and I'm looking forward to it being in the next release.

I do have a question about one point, however. In a [comment made by @klonos back on Feb. 24](https://github.com/backdrop/backdrop-issues/issues/4265#issuecomment-590497150), there is an image that includes the reason why the password is too weak, but I'm not seeing any discussion about that aspect, and it is not in the current PR (at least I'm not seeing it demonstrated in the test site).
Was there discussion about this I'm not recognizing, if so, would someone please point me to it?

For the record, I'm in favor of letting the user know why their password is not acceptable before they submit the form. I hate going to a site and creating a password with no indicator of what is required until after I've thought about it, created a password, and submitted the form. That wastes my time and effort. (I even had one site tell me the password didn't meet requirements after the form was submitted, but never told me what the requirements were -- I just had to guess until I got an acceptable password.)
I do not think it is necessary to let them know what is okay about their password, just what is not okay.

**Comment by indigoxela** [comment]
> there is an image that includes the reason why the password is too weak, but I'm not seeing any discussion about that aspect

The reason for why the password is too weak gets shown in the form error, if submitting although the indicator shows "too weak".

The image provided by @klonos shows functionality I couldn't implement - I wouldn't know how. If you do, a follow-up Issue and PR would be cool.
Personally, I do think that the existing indicator is enough to give users an impression, how strong/weak their password is as they type or copy/paste it  in.

For people who are searching for even stricter constraints, there's a module: https://github.com/backdrop-contrib/passphrase_policy

Again: this issue is about having (configurable) _basic constraints_ in core.

**Comment by oadaeh** [comment]
I wasn't arguing for more constraints, just better communication about the existing ones.
I am aware of that module, but in my opinion, it is too strict.

**Comment by indigoxela** [comment]
@oadaeh Is the current info really that bad?

![pass-constraints](https://user-images.githubusercontent.com/842066/79723427-066b3680-82e6-11ea-8389-9e3a74ef9f5c.png)

Note: in this example all 3 constraints are active, so the three lines directly beneath the password field are the maximum to get shown. If some of the constraints are disabled by admins, the text won't even show.

**Comment by oadaeh** [comment]
Hmm, I don't see the bulleted list. I tried it in Firefox and Chromium (both on Linux). I disabled all privacy Firefox add-ons. I don't have any on Chromium, as I only use it for Google.

![image](https://user-images.githubusercontent.com/376249/79760215-9acc9d80-82d4-11ea-9702-559637106c93.png)

**Comment by olafgrabienski** [comment]
No time to test at the moment, but I notice a difference in the screenshot: The first by @indigoxela is with label "Password" (German: Passwort), the 2nd by @oadaeh with label "**New** password".

**Comment by indigoxela** [comment]
> I don't see the bulleted list.

@oadaeh This is `user/X/edit`, right? This seems to be the one form not having the new help list yet (but has other issues and open PRs).

Try one of these forms:

- Admin adds a new user: admin/people/create
- New user registers: user/register

**Comment by oadaeh** [comment]
Sorry for the delay on replying to this -- I've been preoccupied with other things. I do see the bulleted list on the admin/people/create page, so I have no issues with the functionality. Thanks for that.

**Comment by indigoxela** [comment]
I've updated my PR, rebased and resolved git conflicts...

Ready for a new code review.

@BWPanda did you see my answers to your comments?

**Comment by BWPanda** [comment]
@indigoxela yep, thanks. Still not sure about one of them, perhaps someone else can approve that? Rest were fine.

**Comment by indigoxela** [comment]
@BWPanda Maybe your concern is right, or maybe not. Eventually someone else might have a look.

Previously the indicator comparison was case insensitive, but that baffled me.
`password.toLowerCase() === username.toLowerCase()`

As the password validation (at login time) doesn't consider, for instance, "myusername" and "MyUsErNaMe" the same, it seemed wrong to me that the indicator (when setting a password) does. Sure, it's not the smartest password then, but it's definitely not the same string. Or am I missing something?

**Comment by BWPanda** [comment]
I think the idea is that we don't want people setting passwords that are the same/similar to their username.

Their username might be 'panderson', and even though its technically different, we don't want them setting a password like 'PAnderson' (still too similar and therefore insecure).

**Comment by indigoxela** [comment]
> the idea is that we don't want people setting passwords that are the same/similar to their username

If that's the case, then we have to follow the same logic in php code. :wink: 
I didn't take that into account (yet), as I always read the feature request as "not _same_ as".

So: should we also prevent _alike_ passwords (case insensitive comparison)?

**Comment by klonos** [comment]
>So: should we also prevent _alike_ passwords (case insensitive comparison)?

I have no strong feelings about this, but leaning towards "yes, we should".

PS: I'm also not seeing the bulleted list help text in the PR sandbox.

I have tried:

http://3053.backdrop.backdrop.qa.backdropcms.org/admin/people/create

![Screen Shot 2020-05-17 at 7 32 57 pm](https://user-images.githubusercontent.com/2423362/82154266-4dacfe80-9875-11ea-83f7-f967d965298f.png)

http://3053.backdrop.backdrop.qa.backdropcms.org/user/2/edit

![Screen Shot 2020-05-17 at 7 33 20 pm](https://user-images.githubusercontent.com/2423362/82154272-530a4900-9875-11ea-97f6-d1996c68a1e3.png)

http://3053.backdrop.backdrop.qa.backdropcms.org/user/register (in incognito) only shows username/email, so I cannot test there:

<img width="843" alt="Screen Shot 2020-05-17 at 7 31 20 pm" src="https://user-images.githubusercontent.com/2423362/82154220-06bf0900-9875-11ea-96ae-be5332d1ed8a.png">

**Comment by klonos** [comment]
...I've tried disabling the "Require email verification when a visitor creates an account." checkbox, which states:

> With this setting disabled, users will be logged in immediately upon registering, and may select their own passwords during registration.

...then I was able to see the bulleted help text in http://3053.backdrop.backdrop.qa.backdropcms.org/user/register but not on the same page visit for some reason; so perhaps it's a caching thing 🤷 

Still cannot see it in the user create and user edit forms though.

**Comment by indigoxela** [comment]
> Still cannot see it in the user create and user edit forms though.

The latest merge of upstream, which was a bit tricky anyway, probably broke it. Marking as "needs work", but won't do anything until we really, really get consensus here.

72 Issue comments and ongoing make me think that this won't happen quickly, though.

**Comment by olafgrabienski** [comment]
> Marking as "needs work", but won't do anything until we really, really get consensus here.

Re-reading the issue thread to look for comments that state dissent, I saw a mix of agreement, skepticism, new ideas, testing questions, and more agreement. I guess it's not easy to handle all that feedback! There seems however to be consensus about the feature request in general, and to my mind everytime there were new questions, we've achieved consensus again. Are there open questions where we need more consensus?

**Comment by indigoxela** [comment]
> Are there open questions where we need more consensus?

We should decide whether we prevent _exact_ matches or even _alike_ matches. That's a new request which came in [with this comment](https://github.com/backdrop/backdrop-issues/issues/4265#issuecomment-629770782).

**Comment by olafgrabienski** [comment]
> We should decide whether we prevent *exact* matches or even *alike* matches. That's a new request which came in [with this comment](https://github.com/backdrop/backdrop-issues/issues/4265#issuecomment-629770782).

I see! To make it easy to follow the discussion, I'm pasting the respective statements here:

>>> I think the idea is that we don't want people setting passwords that are the same/similar to their username. (@BWPanda) 

>> I didn't take that into account (yet), as I always read the feature request as "not *same* as". So: should we also prevent *alike* passwords (case insensitive comparison)? (@indigoxela) 

> I have no strong feelings about this, but leaning towards "yes, we should". (@klonos)

I've also no strong feelings about this question, but I'm rather leaning towards "no, we shouldn't". Wanting to prevent passwords, that are *similar* to a username, might open new questions (what is 'similar'? how do we treat similarity in help text?) which complicate things quite a bit while the current approach is already a great step forward. According to a 'small steps at a time philosophy', I'd suggest to implement the request without the similarity approach and to discuss further requests as follow-ups.

**Comment by klonos** [comment]
@indigoxela I would also say that it'd be easier to move this issue forward if the current functionality was left unchanged. That way, there would be no argument over it. The other option is to bring it up during the weekly dev meeting, and see what people think.

We can file a follow-up to change that, and it can either be a task or a bug issue, depending on how people perceive it. All discussion/debate over it can be done in that new issue, to keep scope here clean 😉

**Comment by BWPanda** [comment]
> I would also say that it'd be easier to move this issue forward if the current functionality was left unchanged.

Yes, that's my only concern here - that existing functionality is being changed. If not for that there'd be no issue on my part.

**Comment by herbdool** [comment]
Thank you @indigoxela for all your work on this and for sticking with it! I think we're really close.

**Comment by herbdool** [comment]
I have left a couple comments on the pr.

**Comment by indigoxela** [comment]
Not that I think we have consensus, but I rebased and updated the PR.

Case insensitive comparison is an extra checkbox/feature now.

Several things changed, so this needs complete testing and review again.

**Comment by olafgrabienski** [comment]
I've tested the PR in the sandbox website for the following forms:

- `admin/people/create`
- `user/1/edit`
- `user/register`
  (to see the password field, enable Registration for visitors and disable "Require email verification" on `admin/config/people/settings`)

Works like a charm, not only the strength indicator but also the messages when you submit the form with a weak password.

**Comment by olafgrabienski** [comment]
One small issue, maybe for a follow-up if things get too complicated:

With this PR, the sections on `admin/config/people/settings` are in the following order:

- Registration and cancellation
- URL alias pattern
- **Password constraints**
- Personalization

I'd suggest to move the new "Password constraints" section directly below the related section "Registration and cancellation".

- Registration and cancellation
- **Password constraints**
- URL alias pattern
- Personalization

**Comment by jenlampton** [comment]
I love this improvement! I've added the SR issue label since this is [security related](https://backdropcms.org/core-issue-queue#issue-titles), and tagged it as a milestone candidate incase anyone else agrees :)

**Comment by stpaultim** [comment]
This does look great. Hoping we can get it in next release!

**Comment by klonos** [comment]
I'd also like to see this get in soon 👍 

From memory, the only thing we did not have consensus was re the case sensitivity, which @indigoxela seems to have addressed with the last iteration. All we need is a code review.

I will refrain from nitpicking on this further; if need be, I'll raise follow-ups to improve later 😅

**Comment by klonos** [comment]
I've reviewed this again and left comments in the PR @indigoxela (thanks for all you efforts in getting this in btw 🙏 ...it's a neat security improvement!)

**Comment by indigoxela** [comment]
@klonos: many thanks for reviewing. Consider my answer to one of your comments regarding the logic.

And also: do I really have to rename all those js variables one more time? I feel like I've done that five times already and I don't belief your suggestions are an improvement.

**Comment by indigoxela** [comment]
Hey people, I'm afraid I don't have enough spare time to take care of this issue anymore, so I officially retire from it. Feel free to take over and finalize.

**Comment by klonos** [comment]
@indigoxela it's not the javascript variables that I worry about, since these can easily be changed in the future (they only have effect within that specific .js file); it's the config property names, because these will require an update hook to be changed (because they may be used in various places). I think that for better DX we should have them be something that is self-explanatory, so that we don't need to be adding comments in code to explain what each one is (especially the `no_case` one which is really vague).

**Comment by klonos** [comment]
...oh, I hope that it wasn't my nitpicking that has put you off @indigoxela 😬 😓 ...I'm happy to take over. Although you've done 99.9% of the work, and now I feel like this:

![mr bean - missed a spot](https://user-images.githubusercontent.com/2423362/88527836-789a9800-d006-11ea-97fa-40fb073e42f3.jpg)

**Comment by indigoxela** [comment]
> ...I'm happy to take over

Sure, you're welcome!

**Comment by herbdool** [comment]
If it helps I have permission to update the existing PR and get it over the goalpost (goal line?)

**Comment by klonos** [comment]
Thanks @herbdool 🙏 ...yes, that would also work 👍 

So we need 👀 on the code then.

**Comment by herbdool** [comment]
Ready for testing and review again.

Started off with a small change but I shifted it a bit, based on the comments.

We wanted to keep backwards-compatibility for the password strength indicator. But figured that it would be useful to also allow for the constraints to be used with the indicator even if strong passwords are not enforced. So that meant having to change the UI a bit. I added a checkbox for setting the minimum length as well.

**Comment by klonos** [comment]
We're almost done @herbdool, thanks 🙏 ...last suggestion for consideration: it feels odd to be adding `'#markup' => t('Set constraints'),` to visually group the constraints together; should we change them to `'#type' => 'checkboxes'` instead of having individual `'#type' => 'checkbox'` for each constraint? That way, we'll be able to add `t('Set constraints')` as a label for the checkboxes group. I realize that this means a) reshuffling of the config properties (and you might hit #2557 as well), and b) reordering the checkboxes, so that the `min_length_active` one comes last (so that the `#states`-dependant `min_length` number field comes after it - but in a perfect world, we'd have #2304).

**Comment by herbdool** [comment]
@klonos I tried "checkboxes" but it wasn't working well for me. I can try again, though I think it had issues with one of the elements not being an actual checkbox (min length) and being a child element of the checkboxes. I also tried a fieldset but not sure how people feel about a fieldset within a fieldset.

**Comment by klonos** [comment]
OK, please let me give it a go on my local @herbdool ...if we can't make it work, then sure, let's call this RTBC 😉

**Comment by klonos** [comment]
...just saw that you updated the PR @herbdool, so I'll hold off so that we don't double up on efforts. By the way, in the fresh commit [`0b309cb` (#3053)](https://github.com/backdrop/backdrop/pull/3053/commits/0b309cb9065490347ef1ac7db3ea0d16a4654d8f), once again a checkbox value is saved as string `"0"` instead of numeric `"0"`, and same goes for the value of the number field, which is saved as `"8"` instead of `"8"` ...we should fix that in general 😓

**Comment by herbdool** [comment]
I think it's ready to test again.

**Comment by klonos** [comment]
I did another quick review and testing @herbdool 

- No defaults are set in the PR sandbox (I am assuming that the multiple commits have messed with it, and perhaps refreshing the PR sandbox will fix things):

    ![image](https://user-images.githubusercontent.com/2423362/89081836-efcc8500-d394-11ea-8c06-de0988b396fd.png)

  @quicksketch mentioned during the last dev meeting that perhaps we should not be introducing restrictions by default in existing sites (nor new ones either). So I guess that the default should be to only indicate strength.

- The UI looks and works as expected when clicking around, but you have now inevitably hit the nasty #2732 (see my comment in the PR). Nothing we can do here other than add a custom validation. To test this, try to save the form in this state, and you will see that there is no validation error:

   ![image](https://user-images.githubusercontent.com/2423362/89081961-5356b280-d395-11ea-9fdf-0621811809ec.png)

- Based on the screenshot above, I think that the "Password constraints behavior" radios should be moved after the checkboxes, and that they should be hidden if no constraint has been enabled.

**Comment by jenlampton** [comment]
Why is "enforce strong passwords" not another checkbox under constraints?
It's a constraint, isn't it? Or is it just an indicator?

I don't understand the two radio options. Isn't the first option what it
does now? Does the second option remove the password strength indicator? I
don't think that should be allowed.

**Comment by stpaultim** [comment]
> Why is "enforce strong passwords" not another checkbox under constraints?

I think that the original idea was that an admin started with two options. 

1) Only indicate password strength OR
2) Enforce strong passwords

Then, if and ONLY if an admin selected "enforce strong passwords" they would be presented with options for which password constraints to enforce. 

At that time, if an admin selected "only indicated password strength" - we were going to use the same code that we have now and not offer any constraints.

Since then, it was (as I understand it) suggested that the password constraints would always be available and used to influence the password strength indicator, which in my view, complicates things a little. 

However in my testing, it's not clear to me that the constraint settings are effecting the strength indicator at all. So, I'm not sure why they are visible when the "only indicate password strength" option is selected. 

>  Isn't the first option what it does now?

As far as I can tell, YES.

> Does the second option remove the password strength indicator?

No, I don't think it does (based upon testing). It is only an question of whether or we need to present all the constraints if the admin has selected "Only indicated password strength."

- - - - - - - 

I think that I preferred the version in which the constraints were only visible when enforcing secure passwords. I don't see how it's helpful to have all those checkboxes available all the time. 

BUT - I don't want to hold this issue up any longer. I would just like to be clear about what the intention of the current PR is. 

Should the settings for the password constraints effect the behavior of the strength indicator?

If it that is the current intention and if that is working, I would rather we get this in as it is, rather than risk paralysis. In this case, maybe we just need to improve the wording or help text a bit to address the question from @jenlampton - I am at a loss for ideas right now.

**Comment by jenlampton** [comment]
Thanks for the great explanation Tim, very helpful :)

The PR as is it now needs some work from a UX perspective. I don't
understand the point of the radio button. It should probably be removed if
the checkbox options below are always available now (and I would also
prefer that they always be visible).

With the radio, it may be possible to have conflicting information if you
set the top radio to "do nothing", but then specify something that should
be done in the checkboxes below. (And that's assuming you understand what
the radio options mean)

I love this feature :) why are UIs so hard?

**Comment by herbdool** [comment]
Thanks for feedback everyone. I agree it still needs work. Seems to be a hard UI to figure out partly because there was already two constraints existing: not same as username and case insensitive. So it's a bit odd that we've added a UI to allow to make it weaker.

The other thing is deciding if people get an option to apply constrains in the strength indicator even if someone doesn't want to enforce it. @klonos and started going that direction because it was already applying the username and case already so why not give them option for all of them.

When I have more time I'll reply more.

**Comment by klonos** [comment]
> I love this feature :) why are UIs so hard?

I love this too ❤️ ...I guess that the hard bit is an issue of expectations/perspective 🤷 ...but we have so many brilliant minds making suggestions and providing feedback, that in the end the result is excellent 😉 

>I think that I preferred the version in which the constraints were only visible when enforcing secure passwords. I don't see how it's helpful to have all those checkboxes available all the time.

Here's how I expect this to work:

- Regardless of the indicate/enforce option, password constrains are always "on" when the respective checkbox is ticked, and "off" when unticked; so they should not be hidden. Each constraint influences how the password strength is "graded".
- The "Only indicate password strength" option (with the word "only" being key) is there to advise the user that the password they are entering is weak/strong, but it does not enforce anything. In other words, weak passwords are a possibility: the user account can be created, or the password changed, regardless of password strength.
- The "Enforce strong passwords" is there to throw validation errors, and stop the site admins from creating accounts with weak passwords, or users to change their strong password to a weak one.

Please let me know if the logic above does not make sense, or if you'd expect things to work differently.

>I don't understand the point of the radio button.

Because there are two options: enforce/indicate. If we reduced it to a single checkbox, then users would either need to guess what happens when it is unchecked, or we would need to explain things via a help text. I personally prefer the 2 distinct options over the help text (which more often than not people never read).

>It should probably be removed if the checkbox options below are always available now (and I would also prefer that they always be visible).

The options are always visible - they are always "on" when selected/ticked. The only thing that gets hidden via `#states` now is the "Minimum length" number field (when the min password length checkbox is disabled). I think that this is good UX-wise, as it reduces the noise in the form when the min pass length constrain is disabled.

Perhaps changing the order of the radios to come after the checkboxes would make it easier to understand whats going on:

1. people first select the constraints/factors that influence the password evaluation
2. then we ask them to select what happens when a weak password is set: accept it, or deny submitting the form.

Having written all the above, I'm thinking that "Only indicate password strength" is not that helpful if we are always showing the indicator anyway. Perhaps it would be better as something like "Allow weak passwords"?:

![image](https://user-images.githubusercontent.com/2423362/89200737-1d921380-d5b9-11ea-8cd4-c0a22ce7de7e.png)

**Comment by docwilmot** [comment]
I think it makes sense to always enforce. I cannot imagine that 80% of admins would come to this page and tick those boxes just to see a different JS warning on the password strength indicator.

**Comment by klonos** [comment]
@docwilmot during our previous dev meeting @quicksketch expressed concerns about enforcing password constrains. I think that the intention is to retain current behavior.

**Comment by docwilmot** [comment]
I dont get that. 

If you dont check any boxes, nothing would be enforced. If you check a box, then you *want* it enforced. Whats the problem?

**Comment by herbdool** [comment]
@docwilmot it's not as simple as that. Core already includes two of the constraints and applies them to the password indicator: <> username and case insensitive, as you can see in the JS:

![Screenshot from 2020-08-04 00-03-42](https://user-images.githubusercontent.com/1071483/89252346-724c8300-d5e7-11ea-80c1-ffa6c3f21439.png)

Which is why @klonos and I were thinking why not allow admins to apply the constraints even if not enforcing it. Admittedly the UI still seems to confuse so needs some more work.

**Comment by klonos** [comment]
I was playing on my local, trying to come up with a UI that would be clear. This is what I got so far:

![image](https://user-images.githubusercontent.com/2423362/89294292-f8a8a980-d667-11ea-89b2-99507527be6e.png)

![image](https://user-images.githubusercontent.com/2423362/89294211-dadb4480-d667-11ea-99d8-b54dafa0fdee.png)

**Comment by docwilmot** [comment]
Sorry, still not seeing the issue here. Core already does a notification when your password is weak. In other words core already "allow weak passwords". So this issue should just grant ability to actually enforce the alternative. If the site owner wants to.  n other words again, you dont tick any checkbox. core does what it does now: show a notification. You tick any checkbox, core stops that password if it is unacceptable. In other words we should apply all these constraints to the JS notification by default, but if a user checks a box, enforce that one.

**Comment by klonos** [comment]
> Core already does a notification when your password is weak.

That's right; we're just exposing the settings for that here. The reason we're doing that is because previously we we not rejecting any passwords - now we do, so people may wish to tweak how that works.

> You tick any checkbox, core stops that password if it is unacceptable.

That's what the "reject weak passwords" checkbox does. Are you suggesting something as simple as this instead?:

![image](https://user-images.githubusercontent.com/2423362/89316850-4384ea00-d685-11ea-9f05-2a1863f84dff.png)

If so, then how does a site admin specify at which level of strength passwords are rejected? Do we arbitrarily pick the weak/poor level? For the sake of argument, let's say yes to that. Here's how things would work currently: consider the scenario of `username` as the username, `user@name.com` as email. Then:

- if the password is set to `username` or `USERNAME` or any combination of case, then it is graded as "poor"
- if the password is set to `user@name.com`, it is graded as "good"

So in this scenario, we would be allowing `e-mail == password`, which is one of the initial constraints we set to introduce in this feature. What should the default behavior be then? Should we allow passwords only when they are graded as "excellent" and reject anything bellow that? Or should we tweak the logic so that when `e-mail == password` then that is graded as "poor"?

Finally, are you perhaps suggesting something like this?:

![image](https://user-images.githubusercontent.com/2423362/89321288-70d49680-d68b-11ea-8554-aca188edd131.png)

**Comment by docwilmot** [comment]
> Finally, are you perhaps suggesting something like this?:

Yes!

**Comment by herbdool** [comment]
Here's what I'm thinking now.

Currently Backdrop considers usernames as weak passwords. And it considers usernames as weak while ignoring the case. I feel like we shouldn't even bother offer those as options since that allows people to make it weaker, which is not the intent of this issue.

Then let's take email: let's not offer that as an option either. It's very similar to the username comparison so let's just make it default.

That leaves just password length.

And then I think we should have one checkbox enabled by default:

> Allow users to create weak passwords

Then if an admin unchecked it, the validation will prevent someone from creating a weak password.

The criteria will be the same regardless of whether an admin allows them to create a weak password.

The one thing not mentioned is the complexity calculation that is there currently in the indicator. That would be useful too for the validation. The validation, if enabled, should reject any password that doesn't get a certain score (as is currently calculated in the indicator).

Maybe I'll create a revised pr to demo this, in the next day or so.

**Comment by klonos** [comment]
Thanks @herbdool 🙏  ...I like your reasoning and the overall suggestion, but instead of a ticked-by-default "Allow users to create weak passwords" checkbox, I would prefer a "Reject weak passwords" that's unticked by default. I think that it is closer to an opt-in for the change we are introducing. I would be able to tell better once I've tested the PR sandbox, so I'll wait for that.

In the meantime, and for the sake of being able to compare all suggestions, here's a PR with what I have already drafted on my local (after @docwilmot's last suggestion): https://github.com/backdrop/backdrop/pull/3214

**Comment by docwilmot** [comment]
I like @herbdool's idea of one checkbox.

**Comment by herbdool** [comment]
Here's my simplified version https://github.com/backdrop/backdrop/pull/3215

![Screenshot from 2020-08-06 14-50-27](https://user-images.githubusercontent.com/1071483/89570443-33d7e380-d7f4-11ea-94e4-abc49d9a5fcb.png)

The aim is to add only criteria that will genuinely allow sites to have stronger passwords. So it adds email checking in addition to username. And it leaves the lowercase checking as is. It adds minimum length as an option (since https://blog.codinghorror.com/password-rules-are-bullshit/ says that it is one of the few password rules that is truly helpful). And it allows admins to reject weak passwords (language as per @klonos).

There is one other significant change: in core right now it has a way of calculating the complexity of a password for the indicator. That calculation is in JS. I've added the equivalent function in PHP so that we can also validate with that. So if weak passwords are rejected then any password that is considered "weak" already will be rejected.

**Comment by jenlampton** [comment]
This is fantastic, thanks @herbdool 

I have one recommendation, and that would be to split the requirements in the first paragraph into an item list so that it's easier for people to scan and see the requirements, as follows:

The default criteria for a strong password include checking that it is: 
* not the same as the username 
* not the same as the e-mail address
* sufficiently complex: containing a number of lower- and uppercase letters, digits and symbols. 
* at least as many characters as the minimum number set below

**Comment by herbdool** [comment]
I've made changes suggested by @jenlampton and @klonos. 

And I made a change to use backdrop_strlen() and have it produce same count as using `password.length` in JS (which is UTF-16). Needed for unicode.

**Comment by klonos** [comment]
I agree with @jenlampton re it being more readable as a list instead 👍 ...I've left a code suggestion in the PR to change that text to a `'#type' => 'help'` form element, and to also use a `theme()` function instead. I know that it is currently not styled as the rest of the form help/descriptions, and both @jenlampton and I have come across this recently. We should fix that nevertheless, and once we've fixed the help text element style in the Seven theme, this will inherit it too 😉 

My only comment re this change is this (and I'll just leave it at that, because I don't want to risk getting us into a bike-shedding loop again 😅): ...from a UI/UX perspective, we've removed the checkboxes in order to keep the form less "noisy" and simplify the UI (less scrolling too, which is always a UX++ 👍 ). But we have now basically replaced all the checkboxes with bullet points instead, which occupy the same space (if not more, because there's also the last bullet point, referencing the checkbox immediately after it). So my thinking is that if we are presenting the criteria as bullets, why not have them as options? Meh ...feel free to consider this a venting rant, and ignore me 😌 

Last thing: while playing in my local, I was inspired by the inline selects that we have in `admin/config/people/login`, and have converted the length limit checkbox and number field from this:

![Screen Shot 2020-08-07 at 1 26 46 pm](https://user-images.githubusercontent.com/2423362/89636814-e326c000-d8b1-11ea-91bc-60ea76644200.png)

...to this (`#states` still works, but it enables/disables the min length field instead of showing/hiding it):

![issue-4265-checkbox-number-combo](https://user-images.githubusercontent.com/2423362/89639083-9e048d00-d8b5-11ea-99eb-f7c55d40be4e.gif)

If you guys agree that that is a UX improvement, I've filed a PR against @herbdool's repo: https://github.com/herbdool/backdrop/pull/12

**Comment by herbdool** [comment]
@klonos I've merged in your inline min length PR - I like it.

Re: bullets vs options: the strongest argument against options is that they would only serve to allow someone to make it even weaker than what is already in core, which I don't think anybody wants.

**Comment by herbdool** [comment]
Now we just need a tester and reviewer again so we can update the status of this issue.

**Comment by klonos** [comment]
>I've merged in your inline min length PR - I like it.

Thanks 🙏 ...I've missed one thing while testing on my local though: having the "Password must be at least" bit be a prefix to the number field makes it so that this text cannot be used as a trigger to tick/untick the checkbox. I've added a code suggestion in the PR, which removes the prefix from the number field, and adds it as a label to the checkbox. Things look the same visually, but clicking works as expected now.

**Comment by stpaultim** [comment]
I haven't had a chance to test this, but I really like what I'm seeing in the screenshots! Thanks @herbdool

**Comment by herbdool** [comment]
I wanted to add an aside: I was trying to figure out why the core code was talking about adapting this from Zxcvbn library and ended up finding the Password strength module on d.o. It was a quick fork https://github.com/herbdool/password_strength/issues/1 and is quite similar to what this PR is doing. Perhaps the next step is to actually adapt this contrib module for core. It is still super easy to configure and provides smarter answers. But it does require adding a PHP library to core. Or we can just leave it to contrib, dunno.

**Comment by indigoxela** [comment]
Just one remark: keep in mind what this Issue initially tried to accomplish.

> Enforce some basic password constraints: no username, no email

It seems to me that at least Issue title and description need an update to reflect what's happening now.
And I'm not sure if all people who found basic constraints a cool feature agree that a fully featured password validator (possibly requiring an external library) is the right thing for core.

**Comment by klonos** [comment]
I just run a D7 installation after a long long time 😅 ...here's what the default password input/validator looks out of the box:

![image](https://user-images.githubusercontent.com/2423362/89875208-a6b2d700-dbc5-11ea-9f20-09d1007ef892.png)

As the user types, the constraints/suggestions listed bellow the password get updated (added/removed)

**Comment by herbdool** [comment]
@indigoxela after reflection I agree we likely don't want the external library in core. That can be left for contrib for the time being.

There are a couple things that could improve my PR:

* change the minimum required strength for validation from being hardcoded (currently set to "good") to being set - it would be simple like screenshot above. They would chose from the existing options: weak, fair, good, strong. This would be fairly easy to add.
* add hooks so contrib could adjust the minimum required strength and also the strength that is returned (this was suggested by @quicksketch in the meeting last week too). I'm going to investigate to see how difficult it would be to see if it should be part of this PR or not.

**Comment by herbdool** [comment]
I think it's good enough for this issue. We can have a follow-up for adding another hook. I've added automated tests so should be ready to test

**Comment by herbdool** [comment]
@klonos @stpaultim wondering if we want to try get this in for 1.17 or has enthusiasm waned?

**Comment by indigoxela** [comment]
I'm curious: is it possible to only prevent passwords *same as username* plus a minlength, but do not enforce number/upper/lower/digit/symbol?

**Comment by indigoxela** [comment]
OK, I can answer that myself: partly. I can enforce "not username plus not email" (plus minlength).

It's not really clear (neither as site admin nor as user) what the difference between fair, good and excellent is.
Also, users get no hint, what this mysterious "Please consider making your password more complex" actually means. I had to read the code to figure out.

It could either be longer _or_ contain upper/lower/digit/symbol, both works. Could we add some help text explaining that? Otherwise it gets frustrating.

**Comment by herbdool** [comment]
@indigoxela I've tried to improve the help text.

I think the progress bar provides a lot of feedback on what is a weak, fair, good or excellent password. That's no different from what's there already.

**Comment by herbdool** [comment]
And now I've fixed tests after changing the help text.

**Comment by indigoxela** [comment]
@herbdool many thanks for updating the help text, but that's not actually what I had in mind...

Currently I can enforce strong (=excellent) passwords in two ways:

- setting the minlength to something high
- requiring upper/lower/digit/symbol

If I set the minlength to a high number, I can still set the min strength to disabled or fair, but still enforce strong passwords.

Two examples of _excellent_ passwords (setting minlength to 10, and min strength to excellent):

- `aemai(N1eo.t§2`
- `horse battery set`

An example with minlength 22 and strength _disabled_. - this actually enforces "excellent", too.

- `right horse battery sta` (As you can see, no numbers/symbols)

The hint for users to use punctuation is pretty misleading, I'd rather suggest a hint like:
"The longer, the stronger. If short, include punctuation and numbers." Regarding wording - that's just a suggestion, but I hope you get the point.

Now for the admin form on admin/config/people/settings:

Do we really give enough (and proper) information, what the settings mean? Do admins understand what they enforce by the settings or combinations of settings?

**Comment by indigoxela** [comment]
It seems I'm having trouble to understand settings, too.

I can set a minlength, but then the strength to disabled, and the minlength setting is ignored in validation. That doesn't seem optimal yet.

**Comment by indigoxela** [comment]
Ha! We can solve this easily. It may sound a little weird, but I can explain.

_Drop_ that extra minlength setting in the admin form and also skip the extra evaluation. The function _user_password_evaluate_strength() already considers length for strength and so does the js indicator. Having it as an extra form item makes things unnecessarily complicated.

Then the only thing admins have to decide the is "disabled or fair or good or excellent".

Of course, we have to find the right words for help texts (admin and user).

1. Long passwords are always considered strong (`right horse battery staple`)
2. Shorter passwords have to be more complex (upper/lower/digit/symbol like `f&8#.EkQ9s`)
3. Using username or account email as password is never considered safe
4. (admin only) the higher the "Minimum required password strength" is set, the longer and/or more complex the passwords have to be

Without the extra minlength setting it's much easier to let the js indicator and the php validation behave the same, without making the indicator weaker than it currently is.

@herbdool what do you think? (...Keep on reviewing those other PR, I don't want to distract you. No need to hurry here.)

**Comment by stpaultim** [comment]
I was concerned this issue was getting too complicated. Then I was pleased to see a return to simplicity at this point: https://github.com/backdrop/backdrop-issues/issues/4265#issuecomment-670137214.

The current PR that lets admins select a 'Minimum Required Password Strength' feels like we're getting too complicated again. I think my concern is that these required levels are too ambiguous "fair," "good," and "excellent." If were able to define these for admins, I'd be ok providing this option. 

However, I don't see that this is adding a lot of value. I don't think many admins need this much control. I was totally happy with the earlier simple version. 

Just to be sure, we are looking at this Pr now, right: https://github.com/backdrop/backdrop/pull/3215

**Comment by stpaultim** [comment]
We just had a quick chat about this during our 1.17.0 sprint and am going to put forward a proposal. 

In my view, any time we start to connect password enforcement to the password strength indicator, this issue/PR gets more complicated and less likely to get into core quickly. In my view, the only chance of getting this into 1.17.0 (and that seems unlikely at this point) is if we have some consensus on the following:

1) Implement something like what we were looking at in https://github.com/backdrop/backdrop-issues/issues/4265#issuecomment-670137214. 

At that point, we had a pretty simple option for password constraints that was not tied directly to the password strength indicator. 

We could open up a new issue that tries to improve this and better tie password constraints to the password strength indicator (for future consideration). 

THE OTHER OPTION:

2) We elevate this to the PMC to provide some clear guidelines on what should be in the final PR and wait until 1.18 or later.

- - - - - - - - - - 

I'm just concerned that this issue has over 138 comments and I don't want to loose the work that has been done. Thoughts?

**Comment by quicksketch** [comment]
I like the approach that @herbdool has taken here. It's fairly complex on the PHP side, but as noted, it's just porting what we currently have in JavaScript to PHP so they match in both places. The complexity requirement on the JS side is unchanged.

I agree with everything @indigoxela said 3 comments above (in https://github.com/backdrop/backdrop-issues/issues/4265#issuecomment-683285952).

Keeping the minimum length option could be viable for clarity but it's not working correctly right now (set minimum requirement to "None" and length to `6`, new accounts are still allowed with the password `12345`). Since medium/strong password imply a minimum length anyway, requiring a minimum length is redundant.

So my preference to move this forward is:

- Fix the description text `sufficiently complex: containing a number of lower- and uppercase letters, digits and punctuation` to make clear that length OR variety of characters increases complexity.
- Remove minimum length field option.

In the technical implementation, I'd also add:

- Move the new functions into a dedicated file named `user.password.inc` and include it where necessary.

**Comment by herbdool** [comment]
Alright, seems like consensus is simpler again. I'll make it so. Hopefully I'll get it done before the deadline.

**Comment by herbdool** [comment]
The one thing I'm thinking about is if now we go back to hardcoding the strength level if we revert back to having a simple on/off. I had hardcoded it to 50 (out of 100). 50 is equal to "fair" currently in Backdrop. Hardcoding it doesn't seem right either. But not sure how to reconcile it with what people have brought up.

**Comment by oadaeh** [comment]
If it is hard coded, it couldn't be overridden in contrib.

**Comment by jenlampton** [comment]
I was coming over to this issue to do a code review + testing, but it looks like it may still need work? (changing labels) I'm going to do some testing anyway because I'm excited :) Thank you everyone who's worked on this!

**Comment by jenlampton** [comment]
* I love the password length setting!
<img width="477" alt="Screen Shot 2020-08-30 at 2 13 41 PM" src="https://user-images.githubusercontent.com/397895/91669620-09173d00-eacb-11ea-8401-b6e0c7a53eaa.png">

* I agree with @stpaultim and @quicksketch - I don't understand the `Minimum required password strength` question, and think it should be removed. These values don't have any meaning, and if it's too hard to explain them clearly, we shouldn't be asking the question. 
<img width="1017" alt="Screen Shot 2020-08-30 at 2 14 01 PM" src="https://user-images.githubusercontent.com/397895/91669629-18968600-eacb-11ea-8070-1dc43a24c321.png">

I just tested the [previous PR](https://github.com/backdrop/backdrop/pull/3053) and I prefer that approach. Mainly I appreciate that I understand nearly everything I'm being asked (except what a "strong password" is, exactly).

My overall **favorite approach**, though, is the screenshot @herbdool posted above (posting again here, since that comment was so hard to find :) 
![Screenshot from 2020-08-06 14-50-27](https://user-images.githubusercontent.com/1071483/89570443-33d7e380-d7f4-11ea-94e4-abc49d9a5fcb.png)

The title of this issue is to add constraints specifically for username and email -- and it's not entirely clear we're doing that. Does the indicator show as weak when the username or email is matched in the password? If so, disregard this question, but maybe we should re-title the issue :) 

Maybe thoughts for later on (post Sept 1st) -- I'm not sure either of these are improvements -- but putting them here so I don't forget :)

* The interface above could be simplified even further by removing the `Set minimum password length` checkbox. That text can be used as the label on the integer field, with some description text stating that if left empty, a minimum will not be enforced. 
* Should the password policy be on the "Login settings" page? (admin/config/people/login)

**Comment by herbdool** [comment]
It's ready again for another review.

**Comment by stpaultim** [comment]
From a UI and functionality perspective, this is really good. I think this does what we want it to do in core, but doesn't try to do too much. 

I'll note that from my testing, it appears that it rejects passwords that do not meet the rules that have been set. In fact, the strength indicator does not influence whether or not a password is rejected. I think that is fine, but we should change the language to reflect that. 

For example, if I set a 6 character minimum for passwords, the password "1234567" will be accepted, despite the fact that the indicator says that it's weak. In fact, it is a weak password, but it also does not violate the criteria that we set. I'm ok with that as long as our text is accurate about what is happening. 

I believe that there is a difference between a weak / strong password and an acceptable / unacceptable password. I think it's fine to accept passwords that are weak, but still meet the minimum requirements (while reminding the user that they are in fact weak). 

![Account_settings___PR_3215_backdrop_backdrop_testing_site](https://user-images.githubusercontent.com/3144571/91685118-034b4700-eb1f-11ea-8c19-4c36d45c07a4.png)

I would say something like:

```
- [ ] Reject passwords that do not meet requirements
    Users will not be able to save passwords if they do not meet the above criteria.
```

**Comment by indigoxela** [comment]
I gave testing another stab... but I still find it too confusing.

If I set minlength but do *not* enable "Reject weak passwords", the minlength setting is ignored in php validation, with minlenght `16` I still can use "asdf" as password.

If I set minlength to a high value (24) and enable "Reject weak passwords", the indicator shows "excellent" with a short but complex password, but it still fails validation. Everything's looking fine until I submit the form.

I still belief that it's a bad idea to combine the logic of the indicator (and function _user_password_evaluate_strength()) with _additional_ criteria. This is what makes it so confusing and hard to implement.

**Comment by herbdool** [comment]
@indigoxela I can't replicate: if I set min length to 24 and enter a complex password less than 24 characters it still shows weak to me. Works as expected. This part makes sense to me.
--
Already in core it uses some criteria only to show it's a weak password, but not enforce it so this still just follows the same pattern by extending it to emails and length. And then the option of rejecting altogether.

Maybe it would be good to get fresh eyes on it.

**Comment by herbdool** [comment]
@stpaultim I think you found a bug. Sleepy me missed something last night. I'll fix it this morning.

**Comment by indigoxela** [comment]
> I can't replicate: if I set min length to 24 and enter a complex password less than 24 characters it still shows weak to me. Works as expected.

Neither can I now... strange. So forget about my previous comment. :wink:

**Comment by herbdool** [comment]
@stpaultim fixed. I forgot to update the config on the test site.

**Comment by herbdool** [comment]
Ready for another review. Sorry for that mistake. I've added tests to cover the bug and improved the help text a bit more to reduce confusion when creating passwords.

**Comment by stpaultim** [comment]
@herbdool - OK, that works better. I still do not think it is necessary to tie the strength indicator to the rules so tightly, but you have done so in a way that is now pretty simple and pretty easy to understand (which makes it a good thing).

I am generally pretty happy with look, text and functionality. Will try to test more later to verify that it's all working correctly. 

Please, note the following example:

If I set the minimum length at 6 characters. 

fishfish - is rejected
fishfishfish - is not rejected

I'm totally ok with that, because the second is slightly more secure than the first, just because it's longer. But, it does not have any mixture of "lower- and uppercase letters, digits and punctuation." AGAIN - I am fine with this, but wanted to point it out in case you did not attend to allow this.

**Comment by herbdool** [comment]
@stpaultim thanks! Yes, that is intended that way. We're following the same logic already used for the strength indicator which calculates the complexity by also taking into account the length.

**Comment by stpaultim** [comment]
@herbdool - OK, I found what might be a problem. 

It seems that you can set the minimum characters to 6, but as far as I can tell the system will never accept a 7 letter password. With the minimum set to 6, the following was rejected - 'G3#sh9('

If we will never accept a password of less than 8 characters, we should not allow the user to set the minimum below 8 (or if we do, provide help text to point out the conflict).

**Comment by herbdool** [comment]
@stpaultim I was thinking about this issue just now. since the built-in complexity checker takes into account length (but only in combination) we'd need to know for sure what the minimum is needed - regardless of combo of symbols. Might be safer just to mention it - which it already does - but I've tried not to be too wordy which might overwhelm people.

**Comment by stpaultim** [comment]
I suggest 1 of 2 possible solutions:

1) Change check boxes to radio buttons (have we tried that before?), so user is selecting min length OR "Reject Weak Passwords." 
2) Add a little more help text to "Reject Weak Passwords" option. Something like: `Users will not be able to save passwords if the password strength indicator shows they are weak. This may override the previous setting length setting.`

**Comment by herbdool** [comment]
@stpaultim I think we just tweak the wording. Nothing gets "overridden" really, but hard to explain. It's more like this, if the user types in a password that is:

* their username, they get a failing grade
* their email, they get a failing grade
* shorter than the minimum length, they get a failing grade
* not complex* enough, they get a failing grade (*complex: combination of variety and length)

So it's more like it's a bunch of gates, any of which could slam shut.

**Comment by indigoxela** [comment]
> as far as I can tell the system will never accept a 7 letter password

If that's the case, then we could probably set `'#min' => 8,` in the admin form to prevent lower values.

**Comment by herbdool** [comment]
@indigoxela my thinking too. I've set the minimum. I think that should be enough.

It's probably rare that someone would want to enforce a minimum length but then make it really short. From what I read anything shorter than 10 can be easily cracked nowadays (I'd have to check that number).

It's ready for review/testing again. Hopefully someone will mark it RTBC.

**Comment by stpaultim** [comment]
Right now, it appears that the minimum password length is not working. When I check the minimum password length option at 8 characters but DO NOT check the "reject weak passwords" option, it is still accepting 4-7 character passwords.

**Comment by herbdool** [comment]
@stpaultim that's how it's supposed to work.

It'll only reject weak passwords if you've got it checked.

The min password length is one of the criteria (just like email, username, "complexity"), but it won't be enforced until you've enabled reject weak passwords.

**Comment by herbdool** [comment]
@stpaultim if you think it's good, can you update the "works for me" label?

**Comment by jenlampton** [comment]
> The min password length is one of the criteria (just like email, username, "complexity"), but it won't be enforced until you've enabled reject weak passwords.

This seems counter intuitive. If there's a setting for min password length, it should be respected regardless of any other setting.

If this is the desired behavior, we may be able to fix the problem with a different field label or better description text (after code-freeze). If it were called something like "length for determining a weak password" that might do it.

But, I'd rather have a feature that allowed me to set a minimum password length.

**Comment by stpaultim** [comment]
![password-policy-009](https://user-images.githubusercontent.com/3144571/91911527-e0cc4180-ec76-11ea-9f04-8072f586e509.png)

As a user, I would expect this (the settings in screenshot) to reject passwords shorter than 8 characters and it would feel like a bug to me if it did not. 

If the current behavior is the desired behavior, then (as @jenlampton suggested) I think we need better help text and/or labels to make clear what is happening. 

It's not clear to me that there is much value in providing this option at all, if it will only allow users to set minimum passwords greater than 8 and only have an effect if the "reject weak passwords" box is checked. 

I'm ok with the current behavior, if we fix the language to make it clear that is what is happening. I be even happier if we just removed the minimum length option completely (if it is not an option on it's own).

**Comment by herbdool** [comment]
Both @jenlampton and @stpaultim suggest merging this and tweaking the text after code freeze. Either of you could update the labels if so.

**Comment by jenlampton** [comment]
> I would be even happier if we just removed the minimum length option completely (if it is not an option on it's own).

I agree. I think having a single checkbox for "Reject weak passwords" is enough for 1.17. 

We can add a minimum password length feature later on, since that wasn't originally part of this issue anyway. I think it would be better long-term if there were only a single minimum-length setting. Having two very similar options that need explaining will be more confusing.

**Comment by jenlampton** [comment]
I just did a quick code review, and everything in this PR looks so good! I think it might be simpler to add the additional validation to reject a password less than the minimum length, than to remove the additional setting. It's so nicely integrated to the strength measurement already.

**Comment by jenlampton** [comment]
PR up, and tests still passing :) @stpaultim if you're still around do you want to test this and see if it does what you expect now? https://github.com/backdrop/backdrop/pull/3281

**Comment by herbdool** [comment]
It was my hope that we could have moved towards a framework where new pattern checkers could be added that would improve the strong password checker. (I was inspired by the easy to use module https://www.drupal.org/project/password_strength)

By separating out minimum length, it's no longer really a simple framework IMO and hard to extend in the future.

Aside from that, it makes the hook harder to implement where people could make the requirements only for certain roles, etc.

And it complicates the indicator. Should it still show low strength for short passwords or not?

This last PR seems too last minute. Happening when people like me don't have much time to think about the proposal.

**Comment by jenlampton** [comment]
> And it complicates the indicator. Should it still show low strength for short passwords or not?

The strength calculation is unchanged. Length is still a factor. All the MIN length setting does is prevent people from creating passwords shorter than N characters. It's unrelated to strength.

> This last PR seems too last minute. Happening when people like me don't have much time to think about the proposal.

I'm actually inclined to agree. 

I'd almost prefer we close this issue and start a new one with some very clear goals. The point of the original post here was to "Enforce some basic password constraints: no username, no email" and we've gotten way off track. 

> It was my hope that we could have moved towards a framework where new pattern checkers could be added that would improve the strong password checker.

I saw the new API in the PR and It surprised me (and, I'm not sure it belongs in core). I haven't been following this whole issue obviously, but an API seems like overkill for the original request. 

If we want to add a password strength API why not have a separate issue for that where it can be discussed openly?  The same could be said for the minimum password length setting :) 

This PR has ballooned to include way too many additional features and it's making me uncomfortable to ram it in at the last minute.

**Comment by BWPanda** [comment]
> it's making me uncomfortable to ram it in at the last minute.

Agreed.

**Comment by BWPanda** [comment]
Setting new milestone until we decide what to do here.

**Comment by herbdool** [comment]
I don't really consider adding one hook as adding an "API". It allows people to override one variable nothing more.

A lot of the work on this issue was not last minute at all. It had a lot of back and forth for months, then didn't get any feedback the last 20 days until a flurry the last day. Makes it really hard to get consensus on an issue when it happens like this. It seemed like people were bringing up things which seemed settled ages ago. Frustrating.

**Comment by stpaultim** [comment]
@herbdool - sorry we couldn't make this work. I hope that we take some time to debrief this issue in the near future and discuss the process and how to improve it.

**Comment by jenlampton** [comment]
I'm also sad we couldn't make this work. I know it may be too late for this now, but I'm going back to the comment from 3 hours ago:

> I think having a single checkbox for "Reject weak passwords" is enough for 1.17.

I'm just going to see what it might look like to remove the password length part of this....

Regardless of what we decide to do here, I have created a separate (unrelated) issue to address the minimum password length feature that I wanted in core, so we can remove that whole conversation from this issue. I'm sorry if that was throwing everything off track at the last minute.

**Comment by stpaultim** [comment]
@jenlampton - I don't think the strength indicator is working in your current PR. 

However, the basic UI does appear to address the initial motivation of this issue and does so in a simple way that is easy to understand. Personally, I'd be happy with this. But, it doesn't address the many other issues that came up over the last few months which weren't settled.

Would accepting this PR keep the door open to continue all the other discussions (specifically what @herbdool was trying to do)?

**Comment by herbdool** [comment]
I think this PR is fine for now https://github.com/backdrop/backdrop/pull/3282. I fixed Jen's PR by removing some more of the length variable.

We can adapt later.

**Comment by stpaultim** [comment]
@herbdool - I'm good with this if you are!

**Comment by stpaultim** [comment]
FYI - for anyone just peeking. Here is the current version.

![passwords444359](https://user-images.githubusercontent.com/3144571/91934671-8bf9ec80-ecb1-11ea-8916-810ed9d172ae.png)

**Comment by jenlampton** [comment]
I'd love to get something in, even if it's just a start. This will be a great improvement, and we can continue to add to it as needed. 

We can work on UI text cleanup more in the coming days, but I did re-order the text in the Fieldset on the Account Settings page. This will need feedback in the coming days (Thanks for the screenshot @stpaultim!)

I also changed several instances of `longer and more complex` to `longer or more complex` since either will do. for example: "Please consider making your password longer and more complex".

**Comment by quicksketch** [comment]
Great, I'm glad we could reach a compromise. I'm happy with this and we can refine a bit if needed over the next 2 weeks.

I've merged https://github.com/backdrop/backdrop/pull/3282 into 1.x for 1.17. Thanks so much everyone!

**Comment by herbdool** [comment]
I wonder if we re-added stuff that was intentionally removed in https://github.com/backdrop/backdrop-issues/issues/293, namely the help text was removed. There were a couple good reasons: worked better on small screens, and having a variety of characters, numbers and punctuation doesn't help as much as length.
